File: mass_to_plink/mass_to_plink.py
# coding=utf-8
# 转换基因型数据格式

import os,sys
import pandas as pd

in_excel_name = sys.argv[1] # 第一个参数为Excel文件名
map_file_name = sys.argv[2] # 第二个参数为原始map文件名
out_prefix = sys.argv[3] # 第三个参数为输出的plink文件前缀

# 使用pandas读入 excel
in_file = pd.read_excel(in_excel_name, 'Sheet1')

# 缺失值替换为 "00"
in_file.fillna("00", inplace=True)

# 将第一列和其他列切分开
id_series = in_file.iloc[:,0]
snp_dataframe = in_file.iloc[:,1:]

# 如果基因型为一个字符，则为纯合子，改为两个字符，如 A 改为 AA
row_num = snp_dataframe.shape[0]
col_num = snp_dataframe.shape[1]

for i in range(row_num):
    for j in range(col_num):
        if len(snp_dataframe.iloc[i,j]) == 1: # 判断字符串长度
            snp_dataframe.iloc[i,j] = snp_dataframe.iloc[i,j] * 2

# 生成compound格式的ped格式
# 这里稍微麻烦一点，因为 pandas 合并数据框默认是按照行索引
# 首先值均为0的series，使用 pd.concat() 函数进行拼接

s1 = pd.Series(['0']*row_num)
compound_ped = pd.concat([s1,id_series,s1,s1,s1,s1,snp_dataframe], axis=1)

# 保存为文件，行索引和列标题均不要
compound_ped.to_csv("compound.ped", index=False, header=False, sep=" ")

# 提取map文件
# 首先提取snp名称
snp_list = list(snp_dataframe)

raw_map_file = open(map_file_name,'r')
out_map_file = open("compound.map",'w')

map_dick = {}
for i in raw_map_file:
    f = i.split()
    map_dick[f[1]] = i # snp名称：map文件的整行内容
    
for i in snp_list:
    if i in map_dick:
        out_map_file.write(map_dick[i])
    else:
        print(f"Error: snp {i} in {in_excel_name} but not in {map_file_name}\n")
        sys.exit(1)

raw_map_file.close()
out_map_file.close()

#最后用plink软件转化为正常的plink格式的文件
os.system(f"plink --allow-extra-chr --file compound --compound-genotypes --recode --out {out_prefix} > /dev/null")

File: mass_to_plink/mass_to_plink.R
# 需要设置参数
args<-commandArgs(T)

# 加载 readxl 包，用于读取Excel数据
library(readxl)

# 第一个参数为原始基因型数据的Excel文件名，内容放在 sheet1 中
in_file = read_excel(args[1],sheet = 1)

# 第二个参数为参考map文件，包含了原始基因型数据的所有位点。
map_file = read.table(args[2],header = F)

# 提取原始基因型数据的所有SNP名称
snp_name = names(in_file)[2:length(names(in_file))]

# 向量转为数据框
snp_dataframe = data.frame(snp_name) #生成数据框

# 基因型第一列为芯片号
id_vector = in_file[,1]

# 剔除第一列，即为所有基因型信息的列
data = in_file[,-1] #不要第一列

# 创建一个空的数据框，用于保存处理后的基因型信息
out_data=data.frame()

# 对原始基因型数据进行处理
# Excel 中的空格为缺失，改为00
# 如果只有一个字符，则为纯合子，改为两个字符，如 A 改为 AA
for(i in 1:nrow(data)){
  for (j in 1:ncol(data)){
    if (is.na(data[i,j])){
      out_data[i,j] = "00" #缺失改为00
    } else {
      if (nchar(data[i,j]) == 1){
        out_data[i,j] = paste0(data[i,j],data[i,j])
      } else {
        out_data[i,j] = data[i,j]
      }
    }
  }
}

# out_data 的列名设为空
names(out_data) = NULL

# 生成plink格式的ped文件，使用 cbind 函数合并列，利用了R的广播机制
out_ped = cbind(0,id_vector,0,0,0,0,out_data)

# 生成compound格式的ped文件
write.table(out_ped,file = "compound.ped", col.names = F,quote = F,row.names = F)

# 生成map文件
# 首先对 map_file 和 snp_dataframe 添加列名
names(map_file) = c("chr","snp","3","pos")
names(snp_dataframe) = c("snp") #必须用数据框才能merge

# 使用 merge 函数进行“左连接”, 提取 snp_dataframe 在原始map 文件中的行
# 注意：必须事先保证原始基因型数据的所有SNP均在提供的map文件中。
out_map = merge(snp_dataframe, map_file, by="snp", all.x=TRUE) #顺序错了

# 使用 merge 生成的 out_map 顺序乱了，需要重排列的顺序
new_map = out_map[,names(map_file)] #重排

# 写入文件
write.table(new_map,file = "compound.map", col.names = F,quote = F,row.names = F)

#最后用plink软件转化为正常的plink格式的文件
system("plink --allow-extra-chr --file compound --compound-genotypes --recode --out new > /dev/null")
